llvm/lib/Target/RISCV/RISCVCodeGenPrepare.cpp

```diff
@@ -196,6 +199,42 @@ bool RISCVCodeGenPrepare::expandVPStrideLoad(IntrinsicInst &II) {
   return true;
 }
 
+// The 'fcmp uno/ord/oeq/une/ueq/one/ogt/oge/olt/ole x, 0.0' instructions are
+// equivalent to an FP class test. If the fcmp instruction would be custom
+// lowered or lowered to a libcall, use the is.fpclass intrinsic instead, which
+// is lowered by the back-end without a libcall.
+//
+// This basically reverts the transformations of
+// InstCombinerImpl::foldIntrinsicIsFPClass.
+bool RISCVCodeGenPrepare::visitFCmpInst(FCmpInst &Fcmp) {
+  const auto *TLI = ST->getTargetLowering();
+  const EVT VT = TLI->getValueType(*DL, Fcmp.getOperand(0)->getType());
+  const int ISDOpcode = TLI->InstructionOpcodeToISD(Fcmp.getOpcode());
+
+  auto LegalizeTypeAction = TLI->getTypeAction(Fcmp.getContext(), VT);
+  auto OperationAction = TLI->getOperationAction(ISDOpcode, VT);
+  if ((LegalizeTypeAction != TargetLoweringBase::TypeSoftenFloat &&
+       LegalizeTypeAction != TargetLoweringBase::TypeSoftPromoteHalf) ||
+      OperationAction == TargetLowering::Custom)
+    return false;
```

Comment:
This level of logic really belongs directly in the legalizer

Reply:
Ok, I understand. Moving it to the legalizer. @topperc is it ok to implement it there?

Reply:
Actually the logic in CodeGenPrepare uses `TargetLoweringBase::isFAbsFree`. When I started to implement this, I was wondering if there should be a similar function for `FCmp`, and the whole thing should go into CodegenPrepare instead. Is this a valid approach? Or the right place to do it is in the legalizer.

Reply:
I don't really like having it in codegenprepare in the first place. It really belongs in some combination of DAGCombiner or legalizer, depending on the purpose. The only nice thing is codegenprepare has access to better utilities, like an existing fcmpToClassTest helper and computeKnownFPClass. In principle those should be reimplemented in codegen

Reply:
Ok. Regarding the `InstCombinerImpl::foldIntrinsicIsFPClass`. For a back-end, where lowering the `fcmp` is not cheap, why is it profitable to do this transformation? It is done unconditionally as far as I see it.

Reply:
fcmp is a better canonical form. More code will always understand fcmp than is.fpclass.

> fcmp is not cheap

This is certainly not universally true, and I would say is not the common case. If the target wants something else, that's for the backend to undo for its preferred form.